c7n/resources/ami.py

```diff
@@ -199,6 +287,143 @@ def process_image(self, client, image):
+
+    def validate(self):
+        if self.data.get('remove') == 'matched':
+            found = False
+            for f in self.manager.iter_filters():
+                if isinstance(f, AmiCrossAccountFilter):
+                    found = True
+                    break
+            if not found:
+                raise PolicyValidationError(
+                    "policy:%s filter:%s with matched requires cross-account filter" % (
+                        self.manager.ctx.policy.name, self.type))
+
+    def process(self, images):
+        client = local_session(self.manager.session_factory).client('ec2')
+        for i in images:
+            self.process_image(client, i)
+
+    def process_image(self, client, image):
+        to_add = self.data.get('add')
+        to_remove = self.data.get('remove')
+        # Default is to remove all permissions
+        if not to_add and not to_remove:
+            return client.reset_image_attribute(
+                ImageId=image['ImageId'], Attribute="launchPermission")
+        remove = []
+        add = []
+        account_regex = re.compile('\\d{12}')
+        org_regex = re.compile('arn:[a-zA-Z]+:organizations::\\d{12}:organization/o-.*')
+        ou_regex = re.compile('arn:[a-zA-Z]+:organizations::\\d{12}:ou/o-.*/ou-.*')
+        if to_remove:
+            if 'all' in to_remove:
+                remove.append({'Group': 'all'})
+                to_remove.remove('all')
+            if to_remove == 'matched':
+                to_remove = image.get(AmiCrossAccountFilter.annotation_key)
+            if to_remove:
+                principals = [v for v in to_remove if account_regex.match(v)]
+                if principals:
+                    remove.extend([{'UserId': a} for a in principals])
+                principals = [v for v in to_remove if org_regex.match(v)]
+                if principals:
+                    remove.extend([{'OrganizationArn': a} for a in principals])
+                principals = [v for v in to_remove if ou_regex.match(v)]
+                if principals:
+                    remove.extend([{'OrganizationalUnitArn': a} for a in principals])
+
+        if to_add:
+            if 'all' in to_add:
+                add.append({'Group': 'all'})
+                to_add.remove('all')
+            if to_add:
+                principals = [v for v in to_add if account_regex.match(v)]
+                if principals:
+                    add.extend([{'UserId': a} for a in principals])
+                principals = [v for v in to_add if org_regex.match(v)]
+                if principals:
+                    add.extend([{'OrganizationArn': a} for a in principals])
+                principals = [v for v in to_add if ou_regex.match(v)]
+                if principals:
+                    add.extend([{'OrganizationalUnitArn': a} for a in principals])
+
+        if not remove and not add:
+            return
+        self.manager.retry(client.modify_image_attribute,
```

Comment:
We should be able to merge these two calls, by removing `OperationType` and setting `LaunchPermission={'Remove': remove, 'Add': add}`. From what I can tell we don't lose anything with that change, given that `OperationType` isn't a required parameter.

Reply:
IIRC, the API threw a validation error with both Add and Remove and/or without 'OperationType', but I will double-check that.